File: colcon_ws/src/signal_processing_py/signal_processing_py/fixed_ma_node.py
#!/usr/bin/env python3
"""
Fixed Window Moving Average Processing Node for Signal Processing Pipeline.

This node subscribes to integer and floating-point sensor streams,
applies FIXED WINDOW moving average filters using the pybind11 bindings
of the standalone C++ library, and publishes the filtered results.

Requirements:
- The custom_lib must be built and the Python modules (py_moving_average) available
- ROS2 environment properly sourced

Usage:
    ros2 run signal_processing_py fixed_ma_node
    
    # With parameters
    ros2 run signal_processing_py fixed_ma_node --ros-args -p ma_window_size:=10 -p timeout_seconds:=0.15
"""

import rclpy
from rclpy.node import Node
from std_msgs.msg import Int32, Float32

# Use system-level py_moving_average import (installed via pip in Docker)
# If running locally without pip install, you may need:
#   export PYTHONPATH=/path/to/custom_lib:$PYTHONPATH


class FixedMANode(Node):
    """
    Python ROS2 node that applies FIXED WINDOW moving average filters to sensor streams.
    
    Subscribes to:
        - /encoder_count (Int32): Integer sensor stream
        - /accel_x_mss (Float32): Floating-point sensor stream
        
    Publishes:
        - /fixed_ma_encoder (Int32): Fixed window moving average of integer stream
        - /fixed_ma_accel (Float32): Fixed window moving average of float stream
    """
    
    def __init__(self):
        super().__init__('fixed_ma_node')
        
        # Declare parameters with defaults
        self.declare_parameter('ma_window_size', 5)
        self.declare_parameter('timeout_seconds', 0.15)  # 150ms timeout for dropout gaps
        
        # Get parameter values
        self.ma_window_size = self.get_parameter('ma_window_size').get_parameter_value().integer_value
        self.timeout_seconds = self.get_parameter('timeout_seconds').get_parameter_value().double_value
        
        self.get_logger().info(f"Fixed MA Parameters: window size={self.ma_window_size}, timeout={self.timeout_seconds}s")
        
        # Convert timeout to float (ROS2 returns double by default)
        self.timeout_seconds = float(self.timeout_seconds)
        
        # Import filter modules (will fail if not built)
        self._init_filters()
        
        # Create subscribers
        self.encoder_sub = self.create_subscription(
            Int32, 
            'encoder_count', 
            self.encoder_callback, 
            10
        )
        
        self.accel_sub = self.create_subscription(
            Float32, 
            'accel_x_mss', 
            self.accel_callback, 
            10
        )
        
        # Create publishers
        self.ma_encoder_pub = self.create_publisher(Int32, 'fixed_ma_encoder', 10)
        self.ma_accel_pub = self.create_publisher(Float32, 'fixed_ma_accel', 10)
        
        self.get_logger().info("Fixed MA node initialized")
        self.get_logger().info("Subscribed to: /encoder_count, /accel_x_mss")
        self.get_logger().info("Publishing to: /fixed_ma_encoder, /fixed_ma_accel")
    
    def _init_filters(self):
        """Initialize fixed moving average filter instances with error handling."""
        try:
            # Import and create fixed moving average filters with timeout
            from py_moving_average.FixedMovingAverage.mediumbuffer import Float as MA_Float
            from py_moving_average.FixedMovingAverage.mediumbuffer import Integer as MA_Int
            
            self.ma_encoder = MA_Int(self.ma_window_size, self.timeout_seconds)
            self.ma_accel = MA_Float(self.ma_window_size, self.timeout_seconds)
            
            self.get_logger().info("Fixed moving average filters created with timeout")
            
        except ImportError as e:
            self.get_logger().error(f"Failed to import moving average modules: {e}")
            raise
    
    def encoder_callback(self, msg):
        """Callback for encoder (integer) stream."""
        value = msg.data
        
        # Pass through raw value without filtering for encoder motor topic
        ma_result = float(value)
        
        # Publish raw value to fixed_ma_encoder topic
        self.ma_encoder_pub.publish(Int32(data=int(ma_result)))
        
        # Log occasionally
        if self.ma_encoder.current_size() % 10 == 0:
            self.get_logger().debug(f"Encoder passthrough: raw={value}, published={ma_result:.1f}")
    
    def accel_callback(self, msg):
        """Callback for acceleration (float) stream."""
        value = msg.data
        
        # Apply fixed moving average
        ma_result = self.ma_accel.update(value)
        
        # Publish results
        self.ma_accel_pub.publish(Float32(data=float(ma_result)))
        
        # Log occasionally
        if self.ma_accel.current_size() % 10 == 0:
            self.get_logger().debug(f"Accel fixed MA: raw={value:.3f}, ma={ma_result:.3f}")


def main(args=None):
    rclpy.init(args=args)
    
    try:
        node = FixedMANode()
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    except Exception as e:
        print(f"Error: {e}")
        import traceback
        traceback.print_exc()
    finally:
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: colcon_ws/src/signal_processing_cpp/src/fixed_ma_node.cpp
/**
 * Fixed Window Moving Average Processing Node for Signal Processing Pipeline
 * 
 * This node subscribes to integer and floating-point sensor streams,
 * applies FIXED WINDOW moving average filters using the standalone C++ library,
 * and publishes the filtered results.
 * 
 * Requirements:
 * - The custom_lib must be built and accessible
 * - ROS2 environment properly sourced
 * 
 * Usage:
 *     ros2 run signal_processing_cpp fixed_ma_node
 *     
 *     # With parameters
 *     ros2 run signal_processing_cpp fixed_ma_node --ros-args -p ma_window_size:=10 -p timeout_seconds:=0.15
 */

#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/int32.hpp>
#include <std_msgs/msg/float32.hpp>
#include <chrono>
#include <memory>

// Include the standalone signal processing library headers for moving average
#include "nawe_robotics_lib/running_data/lib/fixed_moving_average.hpp"

using namespace std::chrono_literals;

class FixedMANode : public rclcpp::Node {
public:
    /*! Constructor for FixedMANode
     *  Initialize parameters, create filter, set up subscribers and publishers
     */
    FixedMANode()
        : Node("fixed_ma_node")
    {
        this->declare_parameter<int>("ma_window_size", 5);
        this->declare_parameter<float>("timeout_seconds", 0.15f);  // 150ms timeout for dropout gaps

        int ma_window_size = this->get_parameter("ma_window_size").get_value<int>();
        float timeout_seconds = this->get_parameter("timeout_seconds").get_value<float>();

        RCLCPP_INFO(this->get_logger(), 
                   "Fixed MA Parameters: window size=%d, timeout=%.3fs",
                   ma_window_size, timeout_seconds);

        // Initialize fixed moving average filter for accel with timeout
        // Use buffer size 1000
        ma_accel_ = std::make_unique<FixedMovingAverage<float, 1000>>(ma_window_size, timeout_seconds);

        RCLCPP_INFO(this->get_logger(), "Fixed moving average filter created for accel with timeout");
        RCLCPP_INFO(this->get_logger(), "  Encoder: passthrough (no filtering)");

        encoder_sub_ = this->create_subscription<std_msgs::msg::Int32>(
            "encoder_count", 10, 
            [this](const std_msgs::msg::Int32::SharedPtr msg) {
                this->encoder_callback(msg);
            });

        accel_sub_ = this->create_subscription<std_msgs::msg::Float32>(
            "accel_x_mss", 10, 
            [this](const std_msgs::msg::Float32::SharedPtr msg) {
                this->accel_callback(msg);
            });

        ma_encoder_pub_ = this->create_publisher<std_msgs::msg::Int32>("fixed_ma_encoder", 10);
        ma_accel_pub_ = this->create_publisher<std_msgs::msg::Float32>("fixed_ma_accel", 10);

        RCLCPP_INFO(this->get_logger(), "Fixed MA node initialized");
        RCLCPP_INFO(this->get_logger(), "Subscribed to: /encoder_count, /accel_x_mss");
        RCLCPP_INFO(this->get_logger(), "Publishing to: /fixed_ma_encoder, /fixed_ma_accel");
    }

private:
    /*! Handle encoder data - passthrough without filtering */
    void encoder_callback(const std_msgs::msg::Int32::SharedPtr msg) {
        auto current_time = this->now();
        int32_t value = msg->data;
        
        double dt = (current_time - last_encoder_time_).seconds();
        last_encoder_time_ = current_time;
        
        int32_t ma_result = value;
        
        encoder_update_count_++;
        
        auto ma_msg = std_msgs::msg::Int32();
        ma_msg.data = ma_result;
        ma_encoder_pub_->publish(ma_msg);
        
        if (encoder_update_count_ % 10 == 0) {
            RCLCPP_DEBUG(this->get_logger(), "Encoder passthrough: raw=%d, published=%d, dt=%.3fs",
                         value, ma_result, dt);
        }
    }

    /*! Handle accel data - apply fixed moving average filter */
    void accel_callback(const std_msgs::msg::Float32::SharedPtr msg) {
        auto current_time = this->now();
        float value = msg->data;
        
        double dt = (current_time - last_accel_time_).seconds();
        last_accel_time_ = current_time;
        
        float ma_result = ma_accel_->update(value);
        
        accel_update_count_++;
        
        auto ma_msg = std_msgs::msg::Float32();
        ma_msg.data = ma_result;
        ma_accel_pub_->publish(ma_msg);
        
        if (accel_update_count_ % 10 == 0) {
            RCLCPP_DEBUG(this->get_logger(), "Accel fixed MA: raw=%.3f, ma=%.3f, dt=%.3fs",
                         value, ma_result, dt);
        }
    }

    // Subscribers
    rclcpp::Subscription<std_msgs::msg::Int32>::SharedPtr encoder_sub_;
    rclcpp::Subscription<std_msgs::msg::Float32>::SharedPtr accel_sub_;

    // Publishers
    rclcpp::Publisher<std_msgs::msg::Int32>::SharedPtr ma_encoder_pub_;
    rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr ma_accel_pub_;

    // Filter instances - fixed moving average with buffer size 1000
    std::unique_ptr<FixedMovingAverage<float, 1000>> ma_accel_;
    
    // Timestamps for dt calculation
    rclcpp::Time last_encoder_time_ = this->now();
    rclcpp::Time last_accel_time_ = this->now();
    
    // Counters for occasional logging
    size_t encoder_update_count_ = 0;
    size_t accel_update_count_ = 0;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<FixedMANode>();
    
    rclcpp::spin(node);
    rclcpp::shutdown();
    
    return 0;
}
